h2o-core/src/main/java/hex/ParallelModelBuilder.java

```diff
@@ -0,0 +1,136 @@
+package hex;
+
+import jsr166y.ForkJoinTask;
+import water.util.IcedAtomicInt;
+import water.util.Log;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Dispatcher for parallel model building. Starts building models every time the run method is invoked.
+ * After each model is finished building, the `modelFinished` method is invoked, which in turn invokes modelFeeder callback.
+ * ModelFeeder receives the model built and can deal with it in any way - e.g. put it into a Grid, or discard it if the resulting model
+ * is a failure. It also has the power to invoke the training of any number of new models. Or stop the parallel model builder,
+ * released the barrier inside.
+ */
+public class ParallelModelBuilder extends ForkJoinTask<ParallelModelBuilder> {
+
+  private BiConsumer<Model, ParallelModelBuilder> _onBuildSuccessCallback;
+  private BiConsumer<ModelBuildFailure, ParallelModelBuilder> _onBuildFailureCallback;
+  public IcedAtomicInt _modelInProgressCounter = new IcedAtomicInt();
+  public AtomicBoolean _completed = new AtomicBoolean(false);
+
+  public ParallelModelBuilder(BiConsumer<Model, ParallelModelBuilder> onBuildSuccessCallback,
+                              BiConsumer<ModelBuildFailure, ParallelModelBuilder> onBuildFailureCallback) {
+    Objects.requireNonNull(onBuildSuccessCallback);
+    _onBuildSuccessCallback = onBuildSuccessCallback;
+    _onBuildFailureCallback = onBuildFailureCallback;
+  }
+
+  /**
+   * Runs given collection of {@link ModelBuilder} in parallel. After each model is finished building,
+   * one of the callbacks (on model failure / on model completion) is called.
+   *
+   * @param modelBuilders An {@link Collection} of {@link ModelBuilder} to execute in parallel.
+   */
+  public void run(final Collection<ModelBuilder> modelBuilders) {
+      for (final ModelBuilder modelBuilder : modelBuilders) {
+        _modelInProgressCounter.incrementAndGet();
+
+        // Set the callbacks
+        final Consumer<Model> onModelSuccessfulBuild = this::onModelSuccessfullBuild;
```

Comment:
Are you sure the callbacks will never be serialized? It would be less than ideal to serialize them because it would require to serialize this whole class.

Reply:
Not now. ParallelModelBuilder is only spawned in the `parallelGridSearch` method and it's scope is lifetime of that method only.

But you have a great point - what about the callbacks in the Driver ? In future, the plan is to abandon chunking for smaller datasets and just do one model per node (or multiple per node). This might mean moving a lot of things locally - would such local callbacks still make sense ? Will we be forced to deprecate the Driver callbacks and use H2O's built in RPC to notify the leader node a model is ready ? 

Is the only purpose of the callback this PR and event-driven nature of the ParallelModelBuilder, which I find much better than "looping threads" ?

Reply:
> Will we [...] use H2O's built in RPC to notify the leader node a model is ready ?

This looks like a good approach even now I think. You broadcast the jobId+status, and the listener has enough info to continue.

Reply:
Regarding the callbacks - I am arguing above in favor of making them a custom interface / abstract class to gain more control of what users can pass here.